src/databricks/labs/ucx/source_code/notebooks/sources.py

```diff
@@ -93,35 +112,157 @@ class NotebookLinter:
                 continue
-            linter = self._linter(cell.language.language)
-            for advice in linter.lint(cell.original_code):
+            if isinstance(cell, PythonCell):
+                tree = self._python_trees[cell]
+                advices = self._python_linter.lint_tree(tree)
+            else:
+                linter = self._linter(cell.language.language)
+                advices = linter.lint(cell.original_code)
+            for advice in advices:
                 yield advice.replace(
                     start_line=advice.start_line + cell.original_offset,
                     end_line=advice.end_line + cell.original_offset,
                 )
 
+    def _load_tree_from_notebook(self, notebook: Notebook, register_trees: bool) -> Iterable[Advice]:
+        for cell in notebook.cells:
+            if isinstance(cell, RunCell):
+                yield from self._load_tree_from_run_cell(cell)
+                continue
+            if isinstance(cell, PythonCell):
+                yield from self._load_tree_from_python_cell(cell, register_trees)
+                continue
+
+    def _load_tree_from_python_cell(self, python_cell: PythonCell, register_trees: bool) -> Iterable[Advice]:
+        try:
+            tree = Tree.normalize_and_parse(python_cell.original_code)
+            if register_trees:
+                self._python_trees[python_cell] = tree
+            yield from self._load_children_from_tree(tree)
+        except AstroidSyntaxError as e:
+            yield Failure('syntax-error', str(e), 0, 0, 0, 0)
+
+    def _load_children_from_tree(self, tree: Tree) -> Iterable[Advice]:
+        assert isinstance(tree.node, Module)
+        # look for child notebooks (and sys.path changes that might affect their loading)
+        base_nodes: list[NodeBase] = []
+        base_nodes.extend(self._list_run_magic_lines(tree))
+        base_nodes.extend(SysPathChange.extract_from_tree(self._session_state, tree))
+        if len(base_nodes) == 0:
+            self._python_linter.append_tree(tree)
+            return
+        # append globals
+        globs = cast(Module, tree.node).globals
+        self._python_linter.append_globals(globs)
+        # need to execute things in intertwined sequence so concat and sort them
+        nodes = list(cast(Module, tree.node).body)
+        base_nodes = sorted(base_nodes, key=lambda node: (node.node.lineno, node.node.col_offset))
+        yield from self._load_children_with_base_nodes(nodes, base_nodes)
+        # append remaining nodes
+        self._python_linter.append_nodes(nodes)
+
+    @staticmethod
+    def _list_run_magic_lines(tree: Tree) -> list[MagicLine]:
+
+        def _ignore_problem(_code: str, _message: str, _node: NodeNG) -> None:
+            return None
+
+        commands, _ = MagicLine.extract_from_tree(tree, _ignore_problem)
+        return list(filter(lambda cmd: isinstance(cmd.as_magic(), RunCommand), commands))
```

Comment:
rewrite as for loop, as it'll be difficult to debug once we start getting bug reports about this portion of code

Reply:
Done